### ISSUE TITLE
Feature or doc on programmatically creating correct ConnectionProvider with TenantConnectionResolver

### ISSUE THREAD

**Issue by svinther** [quarkusio/quarkus#11729]
**Description**
When database tenants are completely unknown before "request-time" there is a nice feature in Quarkus to implement a custom `io.quarkus.hibernate.orm.runtime.tenant.TenantConnectionResolver`

This is briefly mentioned at the bottom of this guide:
https://quarkus.io/guides/hibernate-orm#programmatically-resolving-tenants-connections

It would be very helpful with a little more tooling to correctly initialize a custom Datasource, so that the Datasource connections will be correctly entangled with transactions and other parts of the eco-system. 

Since the datasource tenants configuration properties are not to be found in `application.properties` the default implementation `io.quarkus.hibernate.orm.runtime.tenant.DataSourceTenantConnectionResolver` can not be used.

I have tried to init the datasources like in this testproject, with test cases demonstrating that transactions are not working:
https://github.com/svinther/quarkus-problem-h8mt/blob/master/src/main/java/org/acme/getting/started/TenantConnections.java#L35-L55

I would be happy to supply a PR for the guide mentioned above, in case someone can provide a hint enabling me to figure out how to create the Datasources correctly.

**Implementation ideas**

Corresponding to the required datasource properties that goes to `application.properties` :
```
quarkus.datasource.TENANTID.db-kind
quarkus.datasource.TENANTID.username
quarkus.datasource.TENANTID.password
quarkus.datasource.TENANTID.jdbc.url
```

There could be a utility for doing something like this:
```
@ApplicationScoped
@Unremovable
public class TenantConnections implements TenantConnectionResolver {

   @Override
    public ConnectionProvider resolve(String tenant) {
         Properties p = new Properties();
         p.setProperty("quarkus.datasource." + tenant + ".db-kind", "XXX");
         ...
         return NewUtility.compliantConnectionProvider(p);
    }
}

```

**Comment by geoand** [comment]
cc @gsmet @Sanne

**Comment by svinther** [comment]
The essence of this is the question on how to programmatically create a first class citizen `DataSource` that works exactly as if it was created using `application.properties`. Same question here I suppose https://github.com/quarkusio/quarkus/issues/7019, but with no good answers. Would be really helpful with an experts comment on this

**Comment by gavinking** [comment]
Is this issue different or the same as #11861? (I never noticed it before.)

**Comment by svinther** [comment]
Different I think, as #11861 seems to suggest changes in architecture/implementation details on how to select a datasource configured elsewhere, e.g in `application.properties` .

This issue is about how to correctly configure a previously completely unknown datasource during requests. For instance by looking up the configuration properties in etcd (by a request provided tenantid). As can be seen in my reproducer, my attempts so far does not integrate the db connections with transactions.

**Comment by Sanne** [comment]
I agree with the high level idea: it needs to be easier to define new datasources. However we need to think carefully about which level of flexibility to expose: for example I don't think we should allow switching the `db-kind`, as we need that to be a build time constant.

For this reason we can't fully support your request of starting datasources "completely unknown" (like you say); but we could allow providing a dynamic map _tenantId -> DB_NAME_ or _tenantId -> SCHEMA_NAME_ ... same for hostname, credentials.

The point that currently multi-tenant datasources are not enlisted with the transactionmanager seems a bug and we should look at it (unrelated).

**Comment by gavinking** [comment]
Well correct me if I'm misunderstanding this, but if the desire to "define new datasources" at runtime is actually arising out of the need to support multitenancy, then it seems to me that my proposal for datasources which are by nature aware of multitenancy is a *much* better way to address the real underlying need.

Am I way off base here?

**Comment by gavinking** [comment]
What I'm saying here is that the actual requirement is to be able to map tenant ids to catalog names and schema names (and, perhaps, JDBC URLs). I think the perceived need to map tenant ids to dynamically-created datasources is arising out of an architectural bug which we should fix.

**Comment by svinther** [comment]
> I agree with the high level idea: it needs to be easier to define new datasources. However we need to think carefully about which level of flexibility to expose: for example I don't think we should allow switching the `db-kind`, as we need that to be a build time constant.
> 
> For this reason we can't fully support your request of starting datasources "completely unknown" (like you say); but we could allow providing a dynamic map _tenantId -> DB_NAME_ or _tenantId -> SCHEMA_NAME_ ... same for hostname, credentials.
> 

What I was aiming for is dynamic/programmatic multitenancy on the schema and database level. I understand that multitenancy on the dbprovider level is a stretch further and the drivers needed must be known on build time.

In my approach to make this work, I enable the postgres driver by defining a "dummy" default datasource in the `application.properties` like shown in the reproducer example https://github.com/svinther/quarkus-problem-h8mt/blob/master/src/main/resources/application.properties#L11-L16 

> The point that currently multi-tenant datasources are not enlisted with the transactionmanager seems a bug and we should look at it (unrelated).

I don't think there is a bug, as it works well for multitenants defined in `application.properties`. I think that I'm just not able to figure out how to implement the logic for defining correctly behaving datasources, as can be seen in the reproducer example (by transactions not working).

**Comment by svinther** [comment]
For the record, I fixed my specific usecase for now, by doing like this https://github.com/svinther/quarkus-problem-h8mt/pull/1

**Comment by piotr-szybicki** [comment]
So Given that I just walked a mile in @svinther shoes I don't think change in the code base is necessary because the scenario that you are talking about (and I had to deal with) is specific to a very narrow group of developers. And we know who we are so we can just spend 10 extra minutes writing our own implementation of the `TenantConnectionResolver`. What I would advocate here is better documentation on how to write that Resolver.

**Comment by svinther** [comment]
After spending a little more than 10 extra minutes, it seem to be running well for now. But I expect it to break later, and need maintenance. Besides that the solution might be flawed, like with this implementation https://github.com/svinther/quarkus-problem-h8mt/pull/1 it later turned out that I needed to implement my own health checks as well. Some of it also seems a bit hacky, like having to configure a dummy default datasource .

I wonder if it would be appropriate to submit a PR for the hibernate quickstarts with this, if so I will be happy to try and create one ?

**Comment by laurentperez** [comment]
@svinther how did you implement the health checks of /q/health ? thru a custom https://quarkus.io/guides/smallrye-health doing DS lookups ?

also : did it break since sept 2020 ? :) 

we're having a very similar scenario of yours and indeed the whole thing seems hacky and ours is creating driver leaks

I understand db info can't sometimes be known at build processor time, it's outside of a closed world build. But properly creating 1st class DS citizen at runtime is still hacky, couldn't the Arc container of QuarkusConnProvider in some way use some factory to get a proper DS + its health check ?

@piotr-szybicki we are part too of this narrow group, I'm not sure we're that alone in this.

**Comment by svinther** [comment]
@laurentperez I did a poc that seemed to be working at the time https://github.com/svinther/quarkus-problem-h8mt/pull/1 

As I recall, healthcheck works, but I might be wrong. 

I am no longer involved in the establishment where I was using this. It worked well, but I assume it will be broken with newer Quarkus versions.

**Comment by svinther** [comment]
> @svinther does your hack still work? I am hitting same issue with ds not joining the transaction

@marcpuca I don't know if it still works with latest Quarkus. Im not using it any longer as I went elsewhere

**Comment by marcpuca** [comment]
Ok thanks
